### PR TITLE
added details logic (untill the viewmodel)

### PR DESCRIPTION
**Pull request by ozBenTik**

### CODE REVIEW

File: core/src/main/java/data/movies/datasource/MoviesLocalDataSource.kt [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-import di.Upcoming
+import di.*
 import javax.inject.Inject
 
 class MoviesLocalDataSource @Inject constructor(
     @Popular val popularStore: MoviesStore,
     @TopRated val topRatedStore: MoviesStore,
     @Upcoming val upcomingStore: MoviesStore,
     @NowPlaying val nowPlayingStore: MoviesStore,
+    @Recommendations val recommendationsStore: MoviesStore,
```
Comment (Ronelg) [inline core/src/main/java/data/movies/datasource/MoviesLocalDataSource.kt]: refactor the rec. store model to contains map movieId -> List<Movie>

File: core/src/main/java/network/MoviesService.kt [hunk 2]
```diff
@@ -19,4 +21,13 @@ interface MoviesService {
     @GET("movie/top_rated")
     fun getTopRated(@Query("page") page: Int): Call<MovieResponse>
 
+    @GET("/movie/{movie_id}/credits")
```
Comment (Ronelg) [inline core/src/main/java/network/MoviesService.kt]: move leading /

File: domain/src/main/java/com/example/domain/movies/iteractors/UpdateRecommendations.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                        result.data.page,
+                        result.data.movieList
+                    )
+                }
+            }
+    }
+
+    data class Params(val movieId: Int, val pageId: Int)
+
+    object Page {
```
Comment (Ronelg) [inline domain/src/main/java/com/example/domain/movies/iteractors/UpdateRecommendations.kt]: no need for paging

File: domain/src/main/java/com/example/domain/movies/observers/ObserveMovieDetails.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.example.core.data.movies.MoviesCatchSource
+import com.example.domain.SubjectInteractor
+import com.example.model.Movie
+import data.movies.MoviesRepository
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import javax.inject.Inject
+
+class ObserveMovieDetails @Inject constructor(
```
Comment (Ronelg) [inline domain/src/main/java/com/example/domain/movies/observers/ObserveMovieDetails.kt:10]: lets refactor this interactor in way it should not get the store type param
find a way to flat all the movies from all stores and return the movie with id

File: model/src/main/java/com/example/model/Movie.kt [hunk 5]
```diff
@@ -14,7 +14,7 @@ data class Movie (
     @SerializedName("genre_ids")
     val genreList: List<Int> = listOf(),
     @SerializedName("id")
-    val id : Int,
+    val id : Int = -1,
```
Comment (Ronelg) [inline model/src/main/java/com/example/model/Movie.kt]: replace default with 0

File: ui-movies/src/main/java/details/DetailsViewState.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package details
+import com.example.model.Actor
+import com.example.model.Movie
+import util.UiMessage
+
+data class DetailsViewState(
+    val movieDetails: Movie = Movie(),
```
Comment (Ronelg) [inline ui-movies/src/main/java/details/DetailsViewState.kt]: use null if empty (movie)

File: ui-movies/src/main/java/details/MovieDetailsViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    private val observeMovieDetails: ObserveMovieDetails,
+    private val dispatchers: AppCoroutineDispatchers,
+) : ViewModel() {
+
+    private val recommendationsLoadingState = ObservableLoadingCounter()
+    private val actorsLoadingState = ObservableLoadingCounter()
+    private val movieLoadingState = ObservableLoadingCounter()
+    private val uiMessageManager = UiMessageManager()
+
+    val state: StateFlow<DetailsViewState> = combine(
```
Comment (Ronelg) [inline ui-movies/src/main/java/details/MovieDetailsViewModel.kt]: missing observers for rec and actors

File: app/src/main/res/menu/menu.xml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <item
+        android:title="movies"
+        android:icon="@drawable/ic_baseline_local_movies_24"
+        android:id="@+id/movies_nav_graph"
+        />
```
Comment (Ronelg) [inline app/src/main/res/menu/menu.xml:9]: format doc

File: core/src/main/java/data/movies/MoviesCatchSource.kt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package com.example.core.data.movies
+
+enum class MoviesCatchSource {
+    POPULAR,
```
Comment (Ronelg) [inline core/src/main/java/data/movies/MoviesCatchSource.kt]: can be removed

**Review by Ronelg — CHANGES_REQUESTED** [review]

File: core/src/main/java/data/movies/datasource/MoviesLocalDataSource.kt (hunk 1 above)
Comment (ozBenTik) [reply core/src/main/java/data/movies/datasource/MoviesLocalDataSource.kt]: ok

File: core/src/main/java/network/MoviesService.kt (hunk 2 above)
Comment (ozBenTik) [reply core/src/main/java/network/MoviesService.kt]: ok

File: domain/src/main/java/com/example/domain/movies/iteractors/UpdateRecommendations.kt (hunk 3 above)
Comment (ozBenTik) [reply domain/src/main/java/com/example/domain/movies/iteractors/UpdateRecommendations.kt]: ok

File: domain/src/main/java/com/example/domain/movies/observers/ObserveMovieDetails.kt (hunk 4 above)
Comment (ozBenTik) [reply domain/src/main/java/com/example/domain/movies/observers/ObserveMovieDetails.kt:10]: ok

File: model/src/main/java/com/example/model/Movie.kt (hunk 5 above)
Comment (ozBenTik) [reply model/src/main/java/com/example/model/Movie.kt]: ok

File: ui-movies/src/main/java/details/DetailsViewState.kt (hunk 6 above)
Comment (ozBenTik) [reply ui-movies/src/main/java/details/DetailsViewState.kt]: ok

File: core/src/main/java/data/movies/MoviesRepository.kt [hunk 10]
```diff
@@ -60,4 +64,45 @@ class MoviesRepository @Inject constructor(
...
+    // ------------- Movie Catch capabilities -------------------
+    fun getMovieFromCatch(movieId: Int) =
+        combine(
+            local.nowPlayingStore.observeEntries(),
+            local.popularStore.observeEntries(),
+            local.upcomingStore.observeEntries(),
+            local.topRatedStore.observeEntries()
+        ) { nowPlaying, popular, upcoming, topRated ->
+            mutableListOf<Movie>()
+                .addFlatLists(
```
Comment (Ronelg) [inline core/src/main/java/data/movies/MoviesRepository.kt]: addFlatLists is redundant
you can use flatten operator

File: core/src/main/java/extensions/ListExt.kt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package com.example.core.extensions
+
+fun <T> MutableList<T>.addFlatLists(vararg collections: Collection<List<T>>): MutableList<T> {
+    return this.apply {
+        collections.forEach {
+            addAll(it.flatten())
+        }
+    }
```
Comment (Ronelg) [inline core/src/main/java/extensions/ListExt.kt]: you dont need this func
you can use flatten operator

File: ui-movies/src/main/java/details/DetailsViewState.kt [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    val movieDetails: Movie? = null,
+    val movieRefreshing: Boolean = false,
+
+    val recommendations: List<Movie> = emptyList(),
+    val recommendationsRefreshing: Boolean = false,
+
+    val actors: List<Actor> = emptyList(),
+    val actorsRefreshing: Boolean = false,
+
+    val message: UiMessage? = null
```
Comment (Ronelg) [inline ui-movies/src/main/java/details/DetailsViewState.kt]: remove the extra spaces between lines

**Review by Ronelg — CHANGES_REQUESTED** [review]

File: core/src/main/java/data/movies/MoviesCatchSource.kt (hunk 9 above)
Comment (ozBenTik) [reply core/src/main/java/data/movies/MoviesCatchSource.kt]: ok

File: core/src/main/java/data/movies/MoviesRepository.kt (hunk 10 above)
Comment (ozBenTik) [reply core/src/main/java/data/movies/MoviesRepository.kt]: ok

File: core/src/main/java/extensions/ListExt.kt (hunk 11 above)
Comment (ozBenTik) [reply core/src/main/java/extensions/ListExt.kt]: ok

File: ui-movies/src/main/java/details/DetailsViewState.kt (hunk 12 above)
Comment (ozBenTik) [reply ui-movies/src/main/java/details/DetailsViewState.kt]: ok

File: ui-movies/src/main/java/details/MovieDetailsViewModel.kt (hunk 7 above)
Comment (ozBenTik) [reply ui-movies/src/main/java/details/MovieDetailsViewModel.kt]: ok

**Pull request merged**